File: ember_ml/backend/numpy/bitwise/shift_ops.py
"""
NumPy shift bitwise operations for ember_ml.

This module provides NumPy implementations of bitwise shift operations
(left_shift, right_shift, rotate_left, rotate_right).
"""

import numpy as np
from typing import Union

# Import NumpyTensor dynamically within functions
# from ember_ml.backend.numpy.tensor import NumpyTensor
from ember_ml.backend.numpy.types import TensorLike

def left_shift(x: TensorLike, shifts: TensorLike) -> np.ndarray:
    """
    Shift the bits of x to the left by shifts positions.

    Args:
        x: Input tensor or compatible type.
        shifts: Number of bits to shift (integer or tensor).

    Returns:
        NumPy array with x shifted left by shifts bits.
    """
    from ember_ml.backend.numpy.tensor import NumpyTensor
    tensor_ops = NumpyTensor()
    x_arr = tensor_ops.convert_to_tensor(x)
    shifts_arr = tensor_ops.convert_to_tensor(shifts)
    # Ensure shifts is integer type
    if not np.issubdtype(shifts_arr.dtype, np.integer):
         shifts_arr = shifts_arr.astype(np.int32)
    return np.left_shift(x_arr, shifts_arr)

def right_shift(x: TensorLike, shifts: TensorLike) -> np.ndarray:
    """
    Shift the bits of x to the right by shifts positions.

    Args:
        x: Input tensor or compatible type.
        shifts: Number of bits to shift (integer or tensor).

    Returns:
        NumPy array with x shifted right by shifts bits.
    """
    from ember_ml.backend.numpy.tensor import NumpyTensor
    tensor_ops = NumpyTensor()
    x_arr = tensor_ops.convert_to_tensor(x)
    shifts_arr = tensor_ops.convert_to_tensor(shifts)
    # Ensure shifts is integer type
    if not np.issubdtype(shifts_arr.dtype, np.integer):
         shifts_arr = shifts_arr.astype(np.int32)
    return np.right_shift(x_arr, shifts_arr)

def rotate_left(x: TensorLike, shifts: TensorLike, bit_width: int = 32) -> np.ndarray:
    """
    Rotate the bits of x to the left by shifts positions.

    Args:
        x: Input tensor or compatible type (must be unsigned integer type).
        shifts: Number of bits to rotate (integer or tensor).
        bit_width: The bit width of the integer type (e.g., 8, 16, 32, 64).

    Returns:
        NumPy array with x rotated left by shifts bits.
    """
    from ember_ml.backend.numpy.tensor import NumpyTensor
    tensor_ops = NumpyTensor()
    x_arr = tensor_ops.convert_to_tensor(x)
    shifts_arr = tensor_ops.convert_to_tensor(shifts)

    # Ensure unsigned integer type for rotation logic
    if not np.issubdtype(x_arr.dtype, np.unsignedinteger):
        # Attempt to cast, but warn or raise if inappropriate?
        # For now, assume user provides appropriate unsigned type or cast happens before.
        # Example cast (might need adjustment based on desired behavior):
        if np.issubdtype(x_arr.dtype, np.signedinteger):
             print(f"Warning: rotate_left received signed integer type {x_arr.dtype}. Casting to unsigned.")
             # Choose appropriate unsigned type based on bit_width or original type
             if x_arr.dtype == np.int8: x_arr = x_arr.astype(np.uint8)
             elif x_arr.dtype == np.int16: x_arr = x_arr.astype(np.uint16)
             elif x_arr.dtype == np.int32: x_arr = x_arr.astype(np.uint32)
             elif x_arr.dtype == np.int64: x_arr = x_arr.astype(np.uint64)
             else: raise TypeError(f"Unsupported signed integer type for rotate_left: {x_arr.dtype}")
        else:
             raise TypeError(f"rotate_left requires an unsigned integer type, got {x_arr.dtype}")


    # Ensure shifts is integer type
    if not np.issubdtype(shifts_arr.dtype, np.integer):
         shifts_arr = shifts_arr.astype(np.int32)

    # Normalize shifts to be within [0, bit_width)
    # Ensure bit_width array has compatible dtype with shifts_arr for remainder
    bit_width_np = np.array(bit_width, dtype=shifts_arr.dtype) # Use different name
    shifts_arr = np.remainder(shifts_arr, bit_width_np)

    # Perform rotation using shifts and bitwise OR
    # Ensure bit_width array has compatible dtype for subtraction
    bit_width_np = np.array(bit_width, dtype=shifts_arr.dtype)
    # Cast shift amounts explicitly to the target unsigned type for safety
    # Ensure right_shift_amount is calculated using integer arithmetic compatible with shifts_arr
    right_shift_amount = np.subtract(bit_width_np, shifts_arr)
    # Cast the final shift amounts to the same type as x_arr before shifting
    right_shift_amount_casted = right_shift_amount.astype(x_arr.dtype)
    shifts_arr_casted = shifts_arr.astype(x_arr.dtype)

    left_part = np.left_shift(x_arr, shifts_arr_casted)
    right_part = np.right_shift(x_arr, right_shift_amount_casted)

    # Ensure final result is the same dtype as input
    return np.bitwise_or(left_part, right_part).astype(x_arr.dtype)


def rotate_right(x: TensorLike, shifts: TensorLike, bit_width: int = 32) -> np.ndarray:
    """
    Rotate the bits of x to the right by shifts positions.

    Args:
        x: Input tensor or compatible type (must be unsigned integer type).
        shifts: Number of bits to rotate (integer or tensor).
        bit_width: The bit width of the integer type (e.g., 8, 16, 32, 64).

    Returns:
        NumPy array with x rotated right by shifts bits.
    """
    from ember_ml.backend.numpy.tensor import NumpyTensor
    tensor_ops = NumpyTensor()
    x_arr = tensor_ops.convert_to_tensor(x)
    shifts_arr = tensor_ops.convert_to_tensor(shifts)

    # Ensure unsigned integer type for rotation logic
    if not np.issubdtype(x_arr.dtype, np.unsignedinteger):
        if np.issubdtype(x_arr.dtype, np.signedinteger):
             print(f"Warning: rotate_right received signed integer type {x_arr.dtype}. Casting to unsigned.")
             if x_arr.dtype == np.int8: x_arr = x_arr.astype(np.uint8) # Correct astype
             elif x_arr.dtype == np.int16: x_arr = x_arr.astype(np.uint16) # Correct astype
             elif x_arr.dtype == np.int32: x_arr = x_arr.astype(np.uint32) # Correct astype
             elif x_arr.dtype == np.int64: x_arr = x_arr.astype(np.uint64) # Correct astype
             else: raise TypeError(f"Unsupported signed integer type for rotate_right: {x_arr.dtype}")
        else:
             raise TypeError(f"rotate_right requires an unsigned integer type, got {x_arr.dtype}")

    # Ensure shifts is integer type
    if not np.issubdtype(shifts_arr.dtype, np.integer):
         shifts_arr = shifts_arr.astype(np.int32)

    # Normalize shifts to be within [0, bit_width)
    # Ensure bit_width array has compatible dtype with shifts_arr for remainder
    bit_width_np = np.array(bit_width, dtype=shifts_arr.dtype) # Use different name
    shifts_arr = np.remainder(shifts_arr, bit_width_np)

    # Perform rotation using shifts and bitwise OR
    # Ensure bit_width array has compatible dtype for subtraction
    bit_width_np = np.array(bit_width, dtype=shifts_arr.dtype)
    # Cast shift amounts explicitly to the target unsigned type for safety
    left_shift_amount = np.subtract(bit_width_np, shifts_arr).astype(x_arr.dtype)
    shifts_arr_casted = shifts_arr.astype(x_arr.dtype) # Cast shifts amount too

    right_part = np.right_shift(x_arr, shifts_arr_casted)
    left_part = np.left_shift(x_arr, left_shift_amount)

    # Ensure final result is the same dtype as input
    return np.bitwise_or(left_part, right_part).astype(x_arr.dtype)

File: src/commonMain/kotlin/ai/solace/ember/tensor/bitwise/ops/ShiftOps.kt
package ai.solace.ember.tensor.bitwise.ops

/**
 * Shift bitwise operations for Ember ML Kotlin.
 * 
 * This module provides Kotlin implementations of bitwise shift operations
 * (left_shift, right_shift, rotate_left, rotate_right).
 */

/**
 * Shift the bits of x to the left by shifts positions.
 *
 * @param x Input value.
 * @param shifts Number of bits to shift.
 * @return x shifted left by shifts bits.
 */
fun leftShift(x: Int, shifts: Int): Int = x shl shifts
fun leftShift(x: Long, shifts: Int): Long = x shl shifts
fun leftShift(x: UInt, shifts: Int): UInt = x shl shifts
fun leftShift(x: ULong, shifts: Int): ULong = x shl shifts

/**
 * Shift the bits of x to the right by shifts positions.
 *
 * @param x Input value.
 * @param shifts Number of bits to shift.
 * @return x shifted right by shifts bits.
 */
fun rightShift(x: Int, shifts: Int): Int = x shr shifts
fun rightShift(x: Long, shifts: Int): Long = x shr shifts
fun rightShift(x: UInt, shifts: Int): UInt = x shr shifts
fun rightShift(x: ULong, shifts: Int): ULong = x shr shifts

/**
 * Rotate the bits of x to the left by shifts positions.
 *
 * @param x Input value (must be unsigned for proper rotation logic).
 * @param shifts Number of bits to rotate.
 * @param bitWidth The bit width of the integer type (8, 16, 32, 64).
 * @return x rotated left by shifts bits.
 */
fun rotateLeft(x: UInt, shifts: Int, bitWidth: Int = 32): UInt {
    val normalizedShifts = shifts % bitWidth
    val mask: UInt = if (bitWidth >= UInt.SIZE_BITS) UInt.MAX_VALUE else (1u shl bitWidth) - 1u
    val value = x and mask
    if (normalizedShifts == 0) return value
    return ((value shl normalizedShifts) or (value shr (bitWidth - normalizedShifts))) and mask
}

fun rotateLeft(x: ULong, shifts: Int, bitWidth: Int = 64): ULong {
    val normalizedShifts = shifts % bitWidth
    val mask: ULong = if (bitWidth >= ULong.SIZE_BITS) ULong.MAX_VALUE else (1uL shl bitWidth) - 1uL
    val value = x and mask
    if (normalizedShifts == 0) return value
    return ((value shl normalizedShifts) or (value shr (bitWidth - normalizedShifts))) and mask
}

fun rotateLeft(x: UByte, shifts: Int, bitWidth: Int = 8): UByte =
    rotateLeft(x.toUInt(), shifts, bitWidth).toUByte()

fun rotateLeft(x: UShort, shifts: Int, bitWidth: Int = 16): UShort =
    rotateLeft(x.toUInt(), shifts, bitWidth).toUShort()

/**
 * Rotate the bits of x to the right by shifts positions.
 *
 * @param x Input value (must be unsigned for proper rotation logic).
 * @param shifts Number of bits to rotate.
 * @param bitWidth The bit width of the integer type (8, 16, 32, 64).
 * @return x rotated right by shifts bits.
 */
fun rotateRight(x: UInt, shifts: Int, bitWidth: Int = 32): UInt {
    val normalizedShifts = shifts % bitWidth
    val mask: UInt = if (bitWidth >= UInt.SIZE_BITS) UInt.MAX_VALUE else (1u shl bitWidth) - 1u
    val value = x and mask
    if (normalizedShifts == 0) return value
    return ((value shr normalizedShifts) or (value shl (bitWidth - normalizedShifts))) and mask
}

fun rotateRight(x: ULong, shifts: Int, bitWidth: Int = 64): ULong {
    val normalizedShifts = shifts % bitWidth
    val mask: ULong = if (bitWidth >= ULong.SIZE_BITS) ULong.MAX_VALUE else (1uL shl bitWidth) - 1uL
    val value = x and mask
    if (normalizedShifts == 0) return value
    return ((value shr normalizedShifts) or (value shl (bitWidth - normalizedShifts))) and mask
}

fun rotateRight(x: UByte, shifts: Int, bitWidth: Int = 8): UByte =
    rotateRight(x.toUInt(), shifts, bitWidth).toUByte()

fun rotateRight(x: UShort, shifts: Int, bitWidth: Int = 16): UShort =
    rotateRight(x.toUInt(), shifts, bitWidth).toUShort()

/**
 * Array operations for left shift.
 *
 * @param x Input array.
 * @param shifts Array of shift amounts or single shift amount.
 * @return Array with element-wise left shift results.
 */
fun leftShift(x: IntArray, shifts: Int): IntArray = IntArray(x.size) { i -> x[i] shl shifts }
fun leftShift(x: IntArray, shifts: IntArray): IntArray {
    require(x.size == shifts.size) { "Arrays must have the same size" }
    return IntArray(x.size) { i -> x[i] shl shifts[i] }
}

fun leftShift(x: LongArray, shifts: Int): LongArray = LongArray(x.size) { i -> x[i] shl shifts }
fun leftShift(x: LongArray, shifts: IntArray): LongArray {
    require(x.size == shifts.size) { "Arrays must have the same size" }
    return LongArray(x.size) { i -> x[i] shl shifts[i] }
}

/**
 * Array operations for right shift.
 *
 * @param x Input array.
 * @param shifts Array of shift amounts or single shift amount.
 * @return Array with element-wise right shift results.
 */
fun rightShift(x: IntArray, shifts: Int): IntArray = IntArray(x.size) { i -> x[i] shr shifts }
fun rightShift(x: IntArray, shifts: IntArray): IntArray {
    require(x.size == shifts.size) { "Arrays must have the same size" }
    return IntArray(x.size) { i -> x[i] shr shifts[i] }
}

fun rightShift(x: LongArray, shifts: Int): LongArray = LongArray(x.size) { i -> x[i] shr shifts }
fun rightShift(x: LongArray, shifts: IntArray): LongArray {
    require(x.size == shifts.size) { "Arrays must have the same size" }
    return LongArray(x.size) { i -> x[i] shr shifts[i] }
}